docs/design/eth1.md

```diff
@@ -0,0 +1,380 @@
+
+##### Exported:
+
+- `FetchPastLogs(ctx context.Context, fromBlock uint64) ([]ethtypes.Log, error)`
+
+Fetches past blocks.
+
+- `StreamLogs(ctx context.Context) <-chan ethtypes.Log`
+
+Streams logs to the returned channel. Subscribes to block notifications and then requests logs with FilterLogs from `previous_processed_block+1` to `block_from_notification`. Supports offset for `block_from_notification` (8 by default). Might be useful: https://github.com/hydroscan/hydroscan-api/blob/master/task/event_subscriber.go.
+
+If streaming fails, it should reconnect to the node and continue streaming.
+
+- `IsReady(ctx context.Context) (bool, error)`
+
+Returns if node eth1 is ready (not syncing)
+
+##### Unexported:
+
+- `connect(ctx context.Context) error`
+
+Connects to the node.
+
+- `reconnect(ctx context.Context)`
+
+Reconnects to the node using exponential backoff.
+
+#### Metrics:
+
+- Malformed events
+- Processed events
+- Last seen block
+
+#### Tests:
+
+TBD
+
+
+### NodeProber
+
+Checks eth clients readiness.
+
+##### Implemented in https://github.com/bloxapp/ssv/pull/948
+
+
+### EventDispatcher
+
+Dispatches event handling. Gets data from Eth1Client, passes the data to the eventBatcher method, which then passes data to subscribers.
+
+Note: Subscriber pattern may be redundant, consider calling EventDataHandler's method(s) from eventBatcher.
+
+#### Has methods:
+
+- `Start(syncOffset uint64) error`
+
+Fetches past logs and passes them to subscribers for processing. Then it indicates that it's ready. Then runs a loop which retrieves data from Eth1Client event stream (StreamLogs) starring from syncOffset and passes it to the `eventBatcher` method. Before retrieval it should ask NodeProber if nodes are ready.
+
+- `WaitUntilReady()`
+
+Waits until past logs and tasks are processed. Blocks until all events and tasks are processed. For waiting for tasks to finish, instead of communicating between modules, consider using an orchestrator because the process should be synchronous. Consider checking eth nodes readiness here
+
+- `Close() error`
+
+Close the loop and event processing
+
+- `AddSubscriber(subscriber EventSubscriber)`
+
+Adds a subscriber for batched events. Consider replacing it with another pattern as we'll have only 1 subscriber
+
+- `ongoingEventBatcher(events <-chan Event)`
+
+Batches ongoing events from channel and streams batches to subscribers (HandleOngoingEventBatch) as `<-chan EventBatch` (EventBatch represents a batch of events for one block)
+
+- `pastEventBatcher(events []Event)`
+
+Batches past events from channel in `[]EventBatch` and passes the batches to subscribers (HandlePastEventBatches)
+
+#### Metrics:
+
+- Is ready
+- Subscriber count
+
+#### Tests:
+
+TBD
+
+
+### EventDataHandler
+
+Gets batched events from EventDispatcher (EventBatcher). Updates SyncOffset (LastProcessedBlock). Notifies about tasks from events.
+
+#### Has methods:
+
+- `HandleOngoingEventBatch(eventBatchCh <-chan EventBatch)`
```

Comment:
consider `HandleOngoingBlockEvents`

Reply:
Take task logic out of this func. Instead of these 2 funcs, have 1 (receives chan). Task is triggered outside of EventDataHandler

Reply:
```go
func HandleEventLog(blockEvents <-chan BlockEvents) {
//    store to db one by one, no tasks processing/triggering here
}
```